cedar-lean/Cedar/Thm/Validation/Validator.lean

```diff
@@ -0,0 +1,401 @@
+/-
+ Copyright 2022-2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+      https://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+-/
+
+import Cedar.Spec
+import Cedar.Data
+import Cedar.Validation
+import Cedar.Thm.Validation.Typechecker
+import Cedar.Thm.Data
+
+namespace Cedar.Thm
+
+open Cedar.Data
+open Cedar.Spec
+open Cedar.Validation
+
+/-- for a single expression, evaluates to a boolean value (or appropriate error) -/
+def OneEvaluatesCorrectly (expr : Expr) (request : Request) (entities : Entities) : Prop :=
+  ∃ (b : Bool), EvaluatesTo expr request entities b
+
+/-- every policy as an expression evaluates to a boolean value or appropriate error -/
+def AllEvaluateCorrectly (policies : Policies) (request : Request) (entities : Entities) : Prop :=
+  ∀ policy : Policy, policy ∈ policies → OneEvaluatesCorrectly policy.toExpr request entities
```

Comment:
Lean lets you inline the membership check in the declaration of the quantified variable, like this:

```
 ∀ policy ∈ policies, OneEvaluatesCorrectly policy.toExpr request entities
 ```